weld/optimizer/transforms/cse.rs

```diff
@@ -0,0 +1,573 @@
+
+use util::SymbolGenerator;
+use std::collections::{HashMap, HashSet};
+use std::collections::hash_map::Entry;
+
+/// Implements the CSE transform.
+pub fn common_subexpression_elimination(expr: &mut Expr) {
+    use super::inliner;
+    if let Lambda { .. } = expr.kind {
+        expr.uniquify().unwrap();
+        Cse::apply(expr);
+        // Put this here instead of in the pass because the expression tree looks strange
+        // without it, even in tests.
+        inliner::inline_let(expr);
+    }
+}
+
+type ScopeDepth = isize;
+
+/// State for the CSE transformation.
+#[derive(Debug)]
+struct Cse {
+    /// Tracks symbol names.
+    sym_gen: SymbolGenerator,
+}
+
+/// Specifies whether an expression should be considered for common subexpression elimination.
+///
+/// We don't apply CSE to primitives such as literals and identifiers, as well as lambdas, since
+/// those can't be assigned to names in Weld.
+///
+/// Expressions which contain a builder type also cannot be CSE'd, since they are special
+/// expressions that can produce side effects (e.g., memory allocations that should not be aliased
+/// between two values).  In practice, only NewBuilder calls would not be candidates for CSE: since
+/// builder operations must be linear, a common subexpression after uniquify involving `merge` or
+/// `for` should never appear in a valid Weld program.
+trait UseCse {
+    fn use_cse(&self) -> bool;
+}
+
+impl UseCse for Expr {
+    fn use_cse(&self) -> bool {
+        if let Let { .. } = self.kind {
+            return true;
+        }
+
+        if self.ty.contains_builder() {
+            return false;
+        }
+
+        match self.kind {
+            Literal(_) | Ident(_) | Lambda { .. } => false,
+            _ => true
+        }
+    }
+}
+
+impl Cse {
+    /// Apply the CSE transformation on the given expression.
+    pub fn apply(expr: &mut Expr) {
+        let mut cse = Cse {
+            sym_gen: SymbolGenerator::from_expression(expr),
+        };
+
+        // Maps expressions to symbol names.
+        let ref mut bindings = HashMap::new();
+        let ref mut aliases = HashMap::new();
+
+        cse.remove_common_subexpressions(expr, bindings, aliases);
+
+        // Convert the bindings map from Expr -> Symbol to Symbol -> Expr.
+        let ref mut bindings = bindings.drain()
+            .map(|(k, v)| (v, k))
+            .collect::<HashMap<Symbol, Expr>>();
+
+        // FIXME: Update bindings list with expressions. This unnecessarily copies the expression -
+        // we can CSE this by choosing the "outermost" alias and using it everywhere.
+        for (sym, alias_list) in aliases.drain() {
+            for alias in alias_list {
+                let expr = bindings.get(&sym).cloned().unwrap();
+                bindings.insert(alias, expr);
+            }
+        }
+
+        let ref scopes =  cse.build_scope_map(expr, bindings);
+        cse.generate_bindings(expr, bindings, &mut HashSet::new(), &mut vec![], scopes);
```

Comment:
Comment here on what `HashSet::new()` and `vec![]` correspond to (`generated` and `stack`)